oak_protobuf_idl_build/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+
+//! This crate allows compiling protobuf services to Rust in `build.rs` scripts.
+
+use anyhow::Context;
+use prost_build::{Method, Service};
+
+/// Compile Rust server code from the services in the provided protobuf file.
+///
+/// Each method in a service definition must have exactly one comment line of the form `//
+/// method_id: 42`, which is used to generate a stable identifier for that method, which is part of
+/// the serialization protocol used for invocation.
+///
+/// For a service called `TestName`, `compile` generates the following objects:
+///
+/// - a struct named `TestNameServer`, which implements the `oak_idl::InvocationHandler` trait,
+///   dispatching each request to the appropriate method on the underlying service implementation.
+/// - a trait named `TestName`, with a method for each method defined in the protobuf service, and
+///   an additional default method named `serve` which returns an instance of `TestNameServer`; the
+///   developer of a service would usually define a concrete struct and manually implement this
+///   trait for it,
+/// - a struct named `TestNameClient`, exposing a method for each method defined in the protobuf
+///   service. This may be used to directly invoke the underlying handler in order to indirectly
+///   invoke methods on the corresponding `Server` object on the other side of the handler.
+/// - a struct named `TestNameAsyncClient`, similar to `TestNameClient` but with async support.
+pub fn compile(filename: &str) {
+    println!("cargo:rerun-if-changed={}", filename);
+    let mut config = prost_build::Config::new();
+    config.service_generator(Box::new(ServiceGenerator {}));
+    config
+        .compile_protos(&[filename], &["."])
+        .expect("could not compile protobuffer schema");
+}
+
+struct ServiceGenerator {}
+
+impl prost_build::ServiceGenerator for ServiceGenerator {
+    fn generate(&mut self, service: Service, buf: &mut String) {
+        *buf += "\n";
+        *buf += &generate_service(&service).expect("could not generate services");
+        *buf += "\n";
+        *buf += &generate_service_client(&service, false).expect("could not generate clients");
+        *buf += "\n";
+        *buf += &generate_service_client(&service, true).expect("could not generate async clients");
+    }
+}
+
+/// Generate the Rust objects from the input [`Service`] instance, corresponding to a `service`
+/// entry.
+fn generate_service(service: &Service) -> anyhow::Result<String> {
+    let server_name = server_name(service);
+    let service_name = service_name(service);
+    let mut lines = Vec::new();
+    lines.extend(vec![
+        format!("pub struct {server_name}<S> {{"),
+        format!("    service: S"),
+        format!("}}"),
+        format!(""),
+        format!("impl <S: {service_name}> ::oak_idl::Handler for {server_name}<S> {{"),
+        format!("    fn invoke(&mut self, request: ::oak_idl::Request) -> Result<::prost::alloc::vec::Vec<u8>, ::oak_idl::Status> {{"),
```

Comment:
If there is a breaking change in `oak_idl` it will be detected only in tests, right? I wonder if there is a way to capture that those changes at compile time.

Reply:
Yes it will only be in tests, these are just strings from the generator point of view; this is similar to macro expansion in Rust, they are only type checked after expansion.